quelpa.el

```diff
@@ -1806,14 +1864,25 @@ nil."
   (when (quelpa-setup-p) ;if init fails we do nothing
     (let* ((quelpa-upgrade-p (if current-prefix-arg t quelpa-upgrade-p)) ;shadow `quelpa-upgrade-p'
            (quelpa-stable-p quelpa-stable-p) ;shadow `quelpa-stable-p'
-           (cache-item (if (symbolp arg) (list arg) arg)))
+           (quelpa-force-upgrade-p quelpa-force-upgrade-p) ;shadow `quelpa-force-upgrade-p'
+           (cache-item (if (symbolp arg) (list arg) arg))
+           version)
       (quelpa-parse-plist plist)
       (quelpa-parse-stable cache-item)
-      (quelpa-package-install arg)
+      (setq version (quelpa-package-install arg))
+      (setq cache-item `(,(car cache-item) ,@(plist-put (cdr cache-item) :version version)))
       (quelpa-update-cache cache-item)))
   (quelpa-shutdown)
   (run-hooks 'quelpa-after-hook))
 
+;; compatibility
+(advice-add #'package-delete :after
```

Comment:
We should avoid adding mandatory, unconditional advice to important functions like this, which is quite frowned upon, especially in MELPA.  It would be better to only do this cache work when something Quelpa-related is called.

Reply:
I'm only doing `:after` advice, which kind of `add-hook` to `package-delete`, that's not fine?
This advice will only loaded when `quelpa` is loaded so we can assume that we're doing related to quelpa from that point right?

Reply:
This advice is being added unconditionally and permanently to `package-delete`.  It will run every time `package-delete` is called, not just when Quelpa calls it.

It's bad form to add advice unconditionally and permanently this way.  We shouldn't do this unless absolutely necessary.

Reply:
Do we have use quelpa to delete packages? The purpose of this is to remove entry from `quelpa-cache` when we delete package, regardless of where. Also, as I note before, this is `:after` advice, which not really affect any thing `package-delete` is doing.
Perhaps I can say this kind of same thing `async` is doing to have asynchronous byte compiling.

Moreover, this is only loaded when Quelpa is loaded (I'm always delay load Quelpa for start up time gains), and we cannot really unloaded Quelpa, so this would be fine, is it?

Reply:
> Do we have use quelpa to delete packages?  The purpose of this is to remove entry from `quelpa-cache` when we delete package, regardless of where. 

No, of course not.  The issue is that this advice runs even when the package being deleted was not installed with Quelpa.  We should avoid running code like this when Quelpa should not even be involved.

> Also, as I note before, this is `:after` advice, which not really affect any thing `package-delete` is doing.

1.  It affects the behavior of Emacs when the function is called, even when Quelpa should not be involved.
2.  As I've said, it's bad form, and typically against MELPA policy, to add advice unconditionally on load like this.  It's against recommended Emacs coding conventions, see https://www.gnu.org/software/emacs/manual/html_node/elisp/Coding-Conventions.html

> Perhaps I can say this kind of same thing `async` is doing to have asynchronous byte compiling.

I'm not sure what you mean.  It's not async--when advised, `package-delete` will not return until the advice also returns.

> Moreover, this is only loaded when Quelpa is loaded (I'm always delay load Quelpa for start up time gains), and we cannot really unloaded Quelpa, so this would be fine, is it?

No, it's not fine.  Not everyone uses deferred loading.  And regardless, after Quelpa is loaded, the advice is permanent.

So, like I've said, we should not do this unless absolutely necessary.

So, can we talk about whether it *is* absolutely necessary?  Can we talk about whether there are other ways to accomplish this?

Reply:
Sure, do you have any suggestion on this?
I think this is necessary for Quelpa to monitor `package-delete` so we will not accidentally reinstall uninstalled package next time we run `quelpa-upgrade-all`. That would be more annoying to user.
Do we have hook into package install/uninstalling? Perhaps we can create custom quelpa hooks here and advising it to activate the hooks when `package-delete` is called instead?

By `async` doing I mean `emacs-async` package, they have advice around `package--compile`
https://github.com/jwiegley/emacs-async/blob/master/async-bytecomp.el#L137

Reply:
@alphapapa, friendly ping.

Reply:
Is it expensive to determine which packages were installed with Quelpa?  Could that be done when `quelpa-upgrade-all` is called?  That would be much preferable to advising `package-delete`.

> By async doing I mean emacs-async package, they have advice around package--compile

Yes, but that advice is only activated by this minor mode: https://github.com/jwiegley/emacs-async/blob/11769c189528738f9b4e3ce3ca19d73863045c45/async-bytecomp.el#L157

Reply:
It's not that expensive to determine which packages were installed with Quelpa, though it's not that accurate.
We can probably use `quelpa-cache` to see which packages are managed by Quelpa, however user can also upgrade those package using normal package upgrade command, that makes the `quelpa-cache` is not that accurate.
An example scenario is that:
- user forks a package, then install that using Quelpa while waiting for PR to be merged.
- after that, user just follow upstream efforts (without having upgrading fork) using normal package upgrade command

I'm personally using that a lot.

Comeback to this discussion, are you suggesting we're still advice to `package-delete` but limit the time only to when the package is installed by Quelpa?
I think we've already done that by only removing package in `quelpa-cache` when `package-delete` is called

Reply:
> Comeback to this discussion, are you suggesting we're still advice to `package-delete` but limit the time only to when the package is installed by Quelpa?

I don't understand your question.

> I think we've already done that by only removing package in `quelpa-cache` when `package-delete` is called

The point remains that we should not be unconditionally advising `package-delete`.  There must be a better way.

Reply:
This issue has been split into another PR (#155), we should continue to discuss over there.

Reply:
Thanks.